### PR TITLE
Request to merge RAC on Docker changes

### PR DESCRIPTION
**Pull request by psaini79**
1)   Added Sample for Applying patch
2) Fixed Typos
3) Added more checks for RAC

### CODE REVIEW

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/18.3.0/crontabEntry [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+#
+# Copyright (c) 1982-2019 Oracle and/or its affiliates. All rights reserved.
+#
+# Since: January, 2019
+# Author: sanjay.singh@oracle.com, paramdeep.saini@oracle.com
+# Description: Delete  a Grid node and add Oracle Database instance.
+#
+# DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+
+* * * * * /tmp/resetFailedUnits.sh
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/18.3.0/crontabEntry]: Can you explain why you're running this cronjob every minute of every day? Seems a bit excessive without context.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/12.2.0.1/crontabEntry [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+#
+# Copyright (c) 1982-2019 Oracle and/or its affiliates. All rights reserved.
+#
+# Since: January, 2019
+# Author: sanjay.singh@oracle.com, paramdeep.saini@oracle.com
+# Description: Delete  a Grid node and add Oracle Database instance.
+#
+# DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+
+* * * * * /tmp/resetFailedUnits.sh
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/12.2.0.1/crontabEntry]: Can you explain why you're running this cronjob every minute of every day? Seems a bit excessive without context.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/18.3.0/setupDB.sh [hunk 3]
```diff
@@ -4,7 +4,7 @@
 # Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
 #
 # Since: January, 2018
-# Author: paramdeep.saini@oracle.com
+# Author: paramdeep.saini@${DB_USER}.com
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/18.3.0/setupDB.sh]: Find and replace needs to be reverted here.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/18.3.0/setupDB.sh [hunk 4]
```diff
@@ -4,7 +4,7 @@
 # Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/18.3.0/setupDB.sh]: This needs to be updated to 2019.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/18.3.0/setupLinuxEnv.sh [hunk 5]
```diff
@@ -18,3 +18,5 @@ mkdir /responsefiles  && \
 chmod ug+x /opt/scripts/startup/*.sh && \
 yum -y install systemd oracle-database-preinstall-18c net-tools which zip unzip tar openssl expect e2fsprogs openssh-server openssh-client vim-minimal passwd which sudo && \
 yum clean all 
+
+#yum -y install systemd oracle-database-preinstall-18c  net-tools ntpd crontab perl gcc hostname  which zip unzip tar openssl expect e2fsprogs openssh-server openssh-client vim-minimal passwd which sudo xorg-x11-server-Xorg xorg-x11-xauth xorg-x11-apps
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/18.3.0/setupLinuxEnv.sh]: Please don't leave commented out commands in a script. This should be removed. Git will remember it for you.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/AddNode.sh [hunk 6]
```diff
@@ -378,15 +378,15 @@ if [ "${stat}" -ne 0 ]; then
...
+print_message "Running SSH setup for $DB_USER user between nodes ${CLUSTER_NODES[@]}"
+cmd='su - $DB_USER -c "$EXPECT $SCRIPT_DIR/$SETUPSSH $DB_USER \"$DB_HOME/oui/prov/resources/scripts\"  \"${CLUSTER_NODES}\"  \"$ORACLE_PASSWORD\""'
 (eval $cmd) &
 ssh_pid=$!
 wait $ssh_pid
 stat=$?
 
 if [ "${stat}" -ne 0 ]; then
-error_exit "ssh setup for Oracle  user failed!, please make sure you have pass the corect password. You need to make sure that password must be same on all the clustered nodes or the nodes set in existing_cls_nodes env variable for $ORACLE_USER user"
+error_exit "ssh setup for Oracle  user failed!, please make sure you have pass the corect password. You need to make sure that password must be same on all the clustered nodes or the nodes set in existing_cls_nodes env variable for $DB_USER user"
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/AddNode.sh]: Minor double space typo here again.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/checkDBStatus.sh [hunk 7]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 # LICENSE UPL 1.0
 #
-# Copyright (c) 1982-2019 Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/checkDBStatus.sh]: This needs to be changed back to 2019.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/crontabEntry [hunk 8]
```diff
@@ -1 +1,11 @@
...
+#
+# Copyright (c) 1982-2019 Oracle and/or its affiliates. All rights reserved.
+#
+# Since: January, 2019
+# Author: sanjay.singh@oracle.com, paramdeep.saini@oracle.com
+# Description: Delete  a Grid node and add Oracle Database instance.
+#
+# DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+
+* * * * * /tmp/resetFailedUnits.sh
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/crontabEntry]: Can you explain why you're running this cronjob every minute of every day? Seems a bit excessive without context.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/enableRAC.sh [hunk 9]
```diff
@@ -1,14 +1,19 @@
 #!/bin/bash
 # LICENSE UPL 1.0
 #
-# Copyright (c) 1982-2019 Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/enableRAC.sh]: Switch back to 2019.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/installDBBinaries.sh [hunk 10]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 # LICENSE UPL 1.0
 #
-# Copyright (c) 1982-2019 Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/installDBBinaries.sh]: Switch back to 2019.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/setPassword.sh [hunk 11]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 # LICENSE UPL 1.0
 #
-# Copyright (c) 1982-2019 Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/setPassword.sh]: Switch back to 2019.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/setupDB.sh [hunk 12]
```diff
@@ -1,10 +1,10 @@
 #!/bin/bash
 # LICENSE UPL 1.0
 #
-# Copyright (c) 1982-2019 Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/setupDB.sh]: Switch back to 2019.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/setupDB.sh [hunk 13]
```diff
@@ -1,10 +1,10 @@
...
 #
 # Since: January, 2018
-# Author: paramdeep.saini@oracle.com
+# Author: paramdeep.saini@${DB_USER}.com
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/setupDB.sh]: Fix the find and replace error.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/setupGrid.sh [hunk 14]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 # LICENSE UPL 1.0
 #
-# Copyright (c) 1982-2019 Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/setupGrid.sh]: Switch back to 2019.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/setupLinuxEnv.sh [hunk 15]
```diff
@@ -16,5 +16,9 @@
 mkdir /asmdisks && \
 mkdir /responsefiles  && \
 chmod ug+x /opt/scripts/startup/*.sh && \
-yum -y install systemd oracle-database-preinstall-19c net-tools which zip unzip tar openssl expect e2fsprogs openssh-server openssh-client vim-minimal passwd which sudo xorg-x11-server-Xorg xorg-x11-xauth xorg-x11-apps && \
+yum -y install systemd oracle-database-preinstall-19c net-tools which zip unzip tar openssl expect e2fsprogs openssh-server openssh-client vim-minimal passwd which sudo && \
 yum clean all 
+
+## Custom install to install xorg, perl, ntpd,crontab and hostname inside the container
+
+#yum -y install systemd oracle-database-preinstall-19c net-tools ntpd crontab perl gcc hostname  which zip unzip tar openssl expect e2fsprogs openssh-server openssh-client vim-minimal passwd which sudo xorg-x11-server-Xorg xorg-x11-xauth xorg-x11-apps
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/19.3.0/setupLinuxEnv.sh]: Please don't leave commented out commands in scripts. Just remove it. If you need to know what it used to be, that's what git's history is for.

File: OracleDatabase/RAC/OracleRealApplicationClusters/samples/README.md [hunk 16]
```diff
@@ -0,0 +1,7 @@
+Examples of how to extend the Oracle Database Docker Image
+================================
+# applypatch
+Example of how to apply a custom patch or Release Update (RU) to the image. You can apply RU or custom patch on Oracle Grid Infrastructure Home or on Oracle RAC DB Home. For details, pelase refer [README.MD of applypatch](./applypatch/README.md).
+
+# Copyright
+Copyright (c) 2014-2016 Oracle and/or its affiliates. All rights reserved.
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/samples/README.md]: This should be 2019.

File: OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/19.3.0/Dockerfile [hunk 17]
```diff
@@ -0,0 +1,66 @@
+# LICENSE UPL 1.0
+#
+# Copyright (c) 1982-2017 Oracle and/or its affiliates. All rights reserved.
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/19.3.0/Dockerfile]: This should be 2019.

File: OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/19.3.0/Dockerfile [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+# Run: 
+#      $ docker build -f Dockerfile.ee -t oracle/database:12.2.0.1-ee-<patch level> . 
+#
+# Pull base image
+# ---------------
+FROM oracle/database-rac-patch:19.3.0
+
+# Maintainer
+# ----------
+MAINTAINER Gerald Venzl <gerald.venzl@oracle.com>
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/19.3.0/Dockerfile]: Please change the maintainer from Gerald to yourself.

File: OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/19.3.0/Dockerfile [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+#
+# Copyright (c) 1982-2017 Oracle and/or its affiliates. All rights reserved.
+#
+# ORACLE DOCKERFILES PROJECT
+# --------------------------
+# This is the Dockerfile for a patched Oracle Database 12c Release 1 Enterprise Edition
+# 
+# REQUIREMETNS FOR THIS IMAGE
+# ----------------------------------
+# The oracle/database:12.2.0.1-ee image has to exist
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/19.3.0/Dockerfile]: This refers to the wrong image.

File: OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/19.3.0/patches/applyGridPatches.sh_opatchauto [hunk 20]
```diff
@@ -0,0 +1,71 @@
+#!/bin/bash
+# LICENSE UPL 1.0
+#
+# Copyright (c) 1982-2017 Oracle and/or its affiliates. All rights reserved.
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/19.3.0/patches/applyGridPatches.sh_opatchauto]: This should be 2019.

File: OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/19.3.0/patches/applyGridPatches.sh_opatchauto [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+# 
+# Since: January, 2017
+# Author: gerald.venzl@oracle.com
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/19.3.0/patches/applyGridPatches.sh_opatchauto]: I would add your name to the Authors list if you've made any modifications to Gerald's source.

File: OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/README.md [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+
+    [oracle@localhost applypatch]# ./buildPatchedDockerImage.sh -h
+    
+    Usage: ./buildPatchedDockerImage.sh -v [version] [-e | -s] -p [patch label]
+    Builds a patched Docker Image for Oracle Database.
+    
+    Parameters:
+       -v: version to build
+           Choose one of: 19.3.0
+       -e: creates a patched image based on 'Enterprise Edition'
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/README.md]: Are these options correct for patching a RAC image?

File: OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/buildPatchedDockerImage.sh [hunk 23]
```diff
@@ -0,0 +1,130 @@
+#!/bin/bash
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/buildPatchedDockerImage.sh]: The UPL header is missing entirely from this file.

File: OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/buildPatchedDockerImage.sh [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+usage() {
+  cat << EOF
+
+Usage: ./buildPatchedDockerImage.sh -v [version] [-e | -s] -p [patch label]
+Builds a patched Docker Image for Oracle Database.
+  
+Parameters:
+   -v: version to build
+       Choose one of: $(for i in $(ls -d */); do echo -n "${i%%/}  "; done)
+   -e: creates a patched image based on 'Enterprise Edition'
```
Comment (Djelibeybi) [inline OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/buildPatchedDockerImage.sh]: Are these the right parameters for patching a RAC image?

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/12.2.0.1/crontabEntry (hunk 2 above)
Comment (psaini79) [reply OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/12.2.0.1/crontabEntry]: @Djelibeybi , Though I copied this file but it is not being used. It is for our internal customer (within Oracle) who wants to monitor failed units of systemd. I copied this file but not using it anywhere.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/18.3.0/crontabEntry (hunk 1 above)
Comment (psaini79) [reply OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/18.3.0/crontabEntry]: Though I copied this file, it is not being used. It is for our internal customer (within Oracle) who wants to monitor failed units of systemd. I copied this file but not using it anywhere. Let me know if you want me to delete this file then I need to host this file internally  may be on orahub.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/12.2.0.1/setupGrid.sh [hunk 25]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 # LICENSE UPL 1.0
 #
-# Copyright (c) 1982-2019 Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
```
[the remark this answers is not in the document]
Comment (psaini79) [reply OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/12.2.0.1/setupGrid.sh]: Typo. I will fix this.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/12.2.0.1/setupDB.sh [hunk 26]
```diff
@@ -4,7 +4,7 @@
 # Copyright (c) 1982-2018 Oracle and/or its affiliates. All rights reserved.
 #
 # Since: January, 2018
-# Author: paramdeep.saini@oracle.com
+# Author: paramdeep.saini@${DB_USER}.com
```
[the remark this answers is not in the document]
Comment (psaini79) [reply OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/12.2.0.1/setupDB.sh]: Will fix this.

File: OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/19.3.0/patches/applyGridPatches.sh_opatchauto (hunk 20 above)
Comment (psaini79) [reply OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/19.3.0/patches/applyGridPatches.sh_opatchauto]: OK will fix this.

File: OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/19.3.0/patches/applyGridPatches.sh_opatchauto (hunk 21 above)
Comment (psaini79) [reply OracleDatabase/RAC/OracleRealApplicationClusters/samples/applypatch/19.3.0/patches/applyGridPatches.sh_opatchauto]: Ok. I will fix this.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/18.3.0/crontabEntry (hunk 1 above)
Comment (Djelibeybi) [reply OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/18.3.0/crontabEntry]: If the file is not used, don't ship it.

File: OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/12.2.0.1/crontabEntry (hunk 2 above)
Comment (Djelibeybi) [reply OracleDatabase/RAC/OracleRealApplicationClusters/dockerfiles/12.2.0.1/crontabEntry]: As above. If it's not needed, don't include it.

**Pull request merged**